framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,531 @@
+        } else {
+            basetype = "TODOTYPE";
+        }
+
+        if (basetype.contains("[")) {
+            String component = basetype.substring(0, basetype.lastIndexOf('['));
+            printArrayComponentTypeAnnotation(aField.type);
+            printWriter.print(component);
+            printWriter.print(" ");
+            basetype = "[]";
+        }
+        printAnnotations(aField.type.tlAnnotationsHere);
+        printWriter.print(" ");
+        printWriter.print(basetype);
+        printWriter.print(" ");
+        printWriter.print(fieldName);
+    }
+
+    /**
+     * Writes out an import statement for each annotation used in an {@link AScene}.
+     *
+     * <p>{@code DefCollector} is a facility in the Annotation File Utilities for determining which
+     * annotations are used in a given AScene. Here, we use that construct to write out the proper
+     * import statements into a stub file.
+     */
+    private class ImportDefCollector extends DefCollector {
+
+        /**
+         * Constructs a new ImportDefCollector, which will run on the given AScene when its {@code
+         * visit} method is called.
+         *
+         * @param scene the scene whose imported annotations should be printed
+         * @throws DefException if the DefCollector does not succeed
+         */
+        ImportDefCollector(AScene scene) throws DefException {
+            super(scene);
+        }
+
+        /**
+         * Write an import statement for a given AnnotationDef. This is only called once per
+         * annotation used in the scene.
+         *
+         * @param d the annotation definition to print an import for
+         */
+        @Override
+        protected void visitAnnotationDef(AnnotationDef d) {
+            if (!d.name.contains("+")) {
+                printWriter.println("import " + d.name + ";");
+            }
+        }
+    }
+
+    /**
+     * For a given class, this prints the hierarchy of outer classes, and returns the number of
+     * curly braces to close with. The classes are printed with appropriate opening curly braces, in
+     * standard Java style. This routine does not attempt to indent them correctly.
+     *
+     * <p>When an inner class is present in an AScene, its name is something like "Outer.Inner".
+     * Writing a stub file with that name would be useless to the stub parser, which expects inner
+     * classes to be properly nested.
+     *
+     * @param basename the simple name of the class (without the package), in Outer$Inner form
+     * @param classname the fully-qualified name of the class in question, so that it can be printed
+     *     as an enum if it is one
+     * @param aClass the AClass representing the classname
+     * @return the number of outer classes within which this class is nested
+     */
+    private int printClassDefinitions(String basename, String classname, AClass aClass) {
+
+        String nameToPrint = basename;
+        String rest = "";
```

Comment:
Please document what this variable is.  I think it's the outer class within which the given class is nested.  Also give the variable a more descriptive name.